### PR TITLE
GH-41579: [C++][Python][Parquet] Support reading/writing key-value metadata from/to ColumnChunkMetaData

### PR DESCRIPTION
**Pull request by clee704**
### Rationale for this change
Parquet standard allows reading/writing key-value metadata from/to ColumnChunkMetaData, but there is no way to do that with Parquet C++.

### What changes are included in this PR?
Support reading/writing key-value metadata from/to ColumnChunkMetaData with Parquet C++ reader/writer. Support reading key-value metadata from ColumnChunkMetaData with pyarrow.parquet.

### Are these changes tested?
Yes, unit tests are added

### Are there any user-facing changes?
Yes.
- Users can read or write key-value metadata for column chunks with Parquet C++.
- Users can read key-value metadata for column chunks with PyArrow.
- parquet-reader tool prints key-value metadata in column chunks when `--print-key-value-metadata` option is used.

* GitHub Issue: #41579

### CODE REVIEW

File: cpp/src/parquet/metadata.cc
Comment (mapleFU) [inline cpp/src/parquet/metadata.cc:1626]: So once `key_value_metadata()` is called, the metadata is generated, and the metadata would be put in `SetKeyValueMetadata`?

Should we just add a "Put" here?

**Review by mapleFU — COMMENTED** [review]
The idea general LGTM

cc @pitrou @emkornfield @wgtmac

**Comment by wgtmac** [comment]
I saw a related issue: https://issues.apache.org/jira/browse/PARQUET-1648. It seems that parquet-mr does not use it yet.

**Comment by mapleFU** [comment]
> It seems that parquet-mr does not use it yet.

Can parquet-mr reads that?

**Comment by wgtmac** [comment]
I don't think so.

https://github.com/apache/parquet-mr/blob/c241170d9bc2cd8415b04e06ecea40ed3d80f64d/parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java#L1596-L1611

**Comment by mapleFU** [comment]
🤔at least this patch is ok and seems other implementation has thrift, it doesn't break the standard..

**Comment by wgtmac** [comment]
I agree. It should not block us from implementing this.

File: cpp/src/parquet/column_writer.cc
Comment (mapleFU) [inline cpp/src/parquet/column_writer.cc:1409]: Should we also check `!close_` here?

File: cpp/src/parquet/metadata.cc
Comment (wgtmac) [inline cpp/src/parquet/metadata.cc:157]: Can we use `void __set_key_value_metadata(const std::vector<KeyValue> & val);` instead?

File: cpp/src/parquet/column_writer.h
Comment (wgtmac) [inline cpp/src/parquet/column_writer.h:189]: I'd rather not to give the freedom to get the mutable object. Instead, we can implement a similar function equivalent to https://github.com/apache/arrow/blob/main/cpp/src/parquet/file_writer.h#L163

**Review by wgtmac — CHANGES_REQUESTED** [review]

File: cpp/src/parquet/column_writer.h
Comment (pitrou) [inline cpp/src/parquet/column_writer.h:32]: Can you please remove this and instead include `arrow/type_fwd.h`?

File: cpp/src/parquet/column_writer.h
Comment (pitrou) [inline cpp/src/parquet/column_writer.h:57]: It's generally better to avoid `using` directives in non-internal `.h` files.

File: cpp/src/parquet/column_writer.h
Comment (pitrou) [reply cpp/src/parquet/column_writer.h:189]: +1.

File: cpp/src/parquet/metadata.h
Comment (pitrou) [inline cpp/src/parquet/metadata.h:460]: Please let's avoid non-const references, as per our C++ conventions. Instead, you could follow the current builder pattern:
```suggestion
  void SetKeyValueMetadata(const std::shared_ptr<const KeyValueMetadata>&);
```

File: cpp/src/parquet/printer.cc
Comment (pitrou) [inline cpp/src/parquet/printer.cc:75]: Seems like it would be easier to just pass the indentation string?
```suggestion
                           std::string_view indent = "") {
```

File: python/pyarrow/_parquet.pyx
Comment (pitrou) [inline python/pyarrow/_parquet.pyx:521]: Should use `pyarrow_wrap_metadata` instead. Something like (untested):
```suggestion
        wrapped = pyarrow_wrap_metadata(self.metadata.key_value_metadata())
        if wrapped is not None:
            return wrapped.to_dict()
        else:
            return wrapped
```

File: python/pyarrow/tests/parquet/test_metadata.py
Comment (pitrou) [inline python/pyarrow/tests/parquet/test_metadata.py:780]: Can you please remove this print() call (which I assume was meant for debugging)?

File: python/pyarrow/tests/parquet/test_metadata.py
Comment (pitrou) [inline python/pyarrow/tests/parquet/test_metadata.py:781]: 1. Can you test the entire dict contents?
2. Can you add a test where the metadata is `None`?

**Review by pitrou — CHANGES_REQUESTED** [review]

**Comment by pitrou** [comment]
@clee704 Could you please submit a test file to https://github.com/apache/parquet-testing instead of adding it in this PR?

**Comment by clee704** [comment]
> @clee704 Could you please submit a test file to https://github.com/apache/parquet-testing instead of adding it in this PR?

Created https://github.com/apache/parquet-testing/pull/49

File: cpp/src/parquet/printer.cc
Comment (clee704) [reply cpp/src/parquet/printer.cc:75]: To mimic the original output, PrintKeyValueMetadata needs to print two different levels of indentation.

File: cpp/src/parquet/column_writer.h
Comment (clee704) [reply cpp/src/parquet/column_writer.h:57]: We have to choose either this or include "parquet/metadata.h" for consistent use of this class

**Comment by arsnyder16** [comment]
@pitrou @clee704 Where does this PR stand? Would really like to see this exposed in the interface

File: cpp/src/parquet/column_writer.cc
Comment (mapleFU) [inline cpp/src/parquet/column_writer.cc:1109]: 🤔 should for not moved in to help some debugging?

File: cpp/src/parquet/column_writer.h
Comment (mapleFU) [reply cpp/src/parquet/column_writer.h:57]: IMO just `::arrow::KeyValueMetadata` is ok here

File: cpp/src/parquet/column_writer.cc
Comment (mapleFU) [inline cpp/src/parquet/column_writer.cc:1415]: nit: should we throw if `key_value_metadata == nullptr`?

**Review by mapleFU — COMMENTED** [review]
Sorry for delaying, general LGTM, would approve if comment fixed

File: cpp/src/parquet/column_writer.cc
Comment (mapleFU) [reply cpp/src/parquet/column_writer.cc:1415]: in file_writer. the code looks like below:

```c++
  void AddKeyValueMetadata(
      const std::shared_ptr<const KeyValueMetadata>& key_value_metadata) override {
    if (key_value_metadata_ == nullptr) {
      key_value_metadata_ = key_value_metadata;
    } else if (key_value_metadata != nullptr) {
      key_value_metadata_ = key_value_metadata_->Merge(*key_value_metadata);
    }
  }
```

Maybe check null or ignore null is both ok for me

Comment (clee704) [reply cpp/src/parquet/column_writer.cc:1415]: I think you're looking at the old code. The code below is the same as file_writer.

File: cpp/src/parquet/column_writer.cc
Comment (clee704) [reply cpp/src/parquet/column_writer.cc]: done

**Review by mapleFU — APPROVED** [review]
Will merge if parquet-testing is merged

**Comment by mapleFU** [comment]
Oh, Python lint failed...

File: cpp/src/parquet/column_writer_test.cc
Comment (pitrou) [inline cpp/src/parquet/column_writer_test.cc:1739]: Should we use `ASSERT_EQ` here so that the test is more readable? Currently it's a bit confusing what's being checked exactly.

File: cpp/src/parquet/column_writer_test.cc
Comment (pitrou) [inline cpp/src/parquet/column_writer_test.cc:1769]: Same question here.

File: cpp/src/parquet/metadata.cc
Comment (pitrou) [inline cpp/src/parquet/metadata.cc:155]: Nit: avoid a copy
```suggestion
    key_value_metadata.emplace_back(std::move(kv_pair));
```

File: cpp/src/parquet/metadata.cc
Comment (pitrou) [inline cpp/src/parquet/metadata.cc:157]:
```suggestion
  metadata.__set_key_value_metadata(std::move(key_value_metadata));
```

File: cpp/src/parquet/metadata.cc
Comment (pitrou) [inline cpp/src/parquet/metadata.cc:148]: We tend to avoid non-const references in the codebase, so I would suggest to use a pointer instead:
```suggestion
void ToThriftKeyValueMetadata(Metadata* metadata, const KeyValueMetadata& source) {
```

File: cpp/src/parquet/metadata.cc
Comment (pitrou) [inline cpp/src/parquet/metadata.cc:142]: Note that this can probably be made slightly faster if you first build vectors of keys and values, then create the `KeyValueMetadata` from those.

File: cpp/src/parquet/printer.cc
Comment (pitrou) [inline cpp/src/parquet/printer.cc:67]: Can you give this a more distinctive name, such as `PutChars`?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks @clee704! This looks good on the principle, here are some comments.

File: cpp/src/parquet/metadata.cc
Comment (clee704) [reply cpp/src/parquet/metadata.cc:157]: @mapleFU suggested removing the move, so did I. What's the consensus?

File: cpp/src/parquet/column_writer_test.cc
Comment (clee704) [reply cpp/src/parquet/column_writer_test.cc:1739]: I'm against using ASSERT_EQ because the direction is reversed, e.g., `ASSERT_EQ(expected, actual)`, `ASSERT_THAT(actual, matcher)`, and it can hurt the readability when you have both ASSERT_EQ and ASSERT_THAT in the same code. If the readability is a concern, we can write `ASSERT_THAT(actual, Eq(expected))`. What do you thin?

File: cpp/src/parquet/metadata.cc
Comment (mapleFU) [reply cpp/src/parquet/metadata.cc:157]: Aha sorry for misleading, previously I suggest not moving element in member when building ( `metadata_` ). We can applying move for temporary thrift objects

**Comment by mapleFU** [comment]
I merged https://github.com/apache/parquet-testing/pull/49 , you can adjust the link here

**Comment by mapleFU** [comment]
Beside a Py Lint has failed:

```
Python Format............................................................Failed
- hook id: flake8
- exit code: 1

python/pyarrow/tests/parquet/test_metadata.py:788:89: E501 line too long (97 > 88 characters)
```

File: cpp/src/parquet/column_writer_test.cc
Comment (pitrou) [reply cpp/src/parquet/column_writer_test.cc:1739]: I'm ok with `ASSERT_THAT(actual, Eq(expected))`

File: cpp/src/parquet/printer.cc
Comment (mapleFU) [reply cpp/src/parquet/printer.cc:67]: done

**Comment by mapleFU** [comment]
@clee704 Sorry for delaying, I've update this patch myself. This patch LGTM and I'm willing to merge it before August 6th. Would you mind re-checking this?

**Comment by mapleFU** [comment]
Would check CPython build

```
 Error compiling Cython file:
  ------------------------------------------------------------
  ...
          return self.metadata.GetColumnIndexLocation().has_value()

      @property
      def metadata(self):
          """Additional metadata as key value pairs (dict[bytes, bytes])."""
          wrapped = pyarrow_wrap_metadata(self.metadata.key_value_metadata())
                    ^
  ------------------------------------------------------------

  pyarrow/_parquet.pyx:514:18: undeclared name not builtin: pyarrow_wrap_metadata

  Error compiling Cython file:
  ------------------------------------------------------------
  ...
          return self.metadata.GetColumnIndexLocation().has_value()

      @property
      def metadata(self):
          """Additional metadata as key value pairs (dict[bytes, bytes])."""
          wrapped = pyarrow_wrap_metadata(self.metadata.key_value_metadata())
                                                                          ^
  ------------------------------------------------------------

  pyarrow/_parquet.pyx:514:72: Cannot convert 'shared_ptr[const CKeyValueMetadata]' to Python object
  [17/136] Compiling Cython CXX source for _dataset...
  performance hint: pyarrow/_dataset.pyx:1947:35: Exception check after calling 'GetResultValue[shared_ptr[CRandomAccessFile]]' will always require the GIL to be acquired. Declare 'GetResultValue[shared_ptr[CRandomAccessFile]]' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.
  performance hint: pyarrow/_dataset.pyx:3343:41: Exception check after calling 'GetResultValue[shared_ptr[CRecordBatch]]' will always require the GIL to be acquired. Declare 'GetResultValue[shared_ptr[CRecordBatch]]' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.
  performance hint: pyarrow/_dataset.pyx:3387:34: Exception check after calling 'GetResultValue[CTaggedRecordBatch]' will always require the GIL to be acquired. Declare 'GetResultValue[CTaggedRecordBatch]' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.
  performance hint: pyarrow/_dataset.pyx:3784:42: Exception check after calling 'GetResultValue[CTaggedRecordBatchIterator]' will always require the GIL to be acquired. Declare 'GetResultValue[CTaggedRecordBatchIterator]' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.
  [18/136] Building CXX object CMakeFiles/arrow_python.dir/pyarrow/src/arrow/python/datetime.cc.o
```

**Comment by mapleFU** [comment]
@wgtmac @pitrou Do you have any more comments here? Lets move forward now

**Review by wgtmac — APPROVED** [review]
+1 on the C++ side

**Comment by mapleFU** [comment]
@pitrou Would you mind take a look at this? I'm planning to merge this pr this week if no negative comments

File: cpp/src/parquet/column_writer_test.cc
Comment (pitrou) [inline cpp/src/parquet/column_writer_test.cc:1733]: Can you add a test where `AddKeyValueMetadata` is called twice?

File: cpp/src/parquet/metadata.cc
Comment (pitrou) [inline cpp/src/parquet/metadata.cc:139]: Perhaps call this `FromThriftKeyValueMetadata`?

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Sorry, I approved while the two comments should probably be addressed before merging this.

**Comment by mapleFU** [comment]
Thanks all, merged!

File: python/pyarrow/tests/parquet/conftest.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/parquet/conftest.py:37]: This is the first time we introduce a pyarrow test that requires this to be set up. Do we want to require that strictly, or should we skip the test if the env variable is not set? 

In any case this caused a failure in one of the nightly crossbow builds which doesn't have this env variable set (python-emscriptem, https://github.com/ursacomputing/crossbow/actions/runs/10463078918/job/28974494645#step:7:15654)

Comment (mapleFU) [reply python/pyarrow/tests/parquet/conftest.py:37]: Both is ok for me but I prefer read that 🤔

**Pull request merged**
